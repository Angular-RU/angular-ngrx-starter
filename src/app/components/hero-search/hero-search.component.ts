import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  OnDestroy
} from '@angular/core';

import { Observable ,  Subject ,  of } from 'rxjs';
import { Store, select } from '@ngrx/store';

import { Hero } from '@appModels/hero';

import * as fromSelectors from '@appStore/selectors';
import * as fromStore from '@appStore/index';

import { SearchReset, Search } from '@appStore/actions/search.actions';

@Component({
  selector: 'app-hero-search',
  templateUrl: './hero-search.component.html',
  styleUrls: ['./hero-search.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class HeroSearchComponent implements OnInit, OnDestroy {
  searchTerm$: Observable<string>;
  heroes$: Observable<Hero[]>;

  constructor(private store: Store<fromStore.State>) {}

  ngOnInit(): void {
    this.heroes$ = this.store.pipe(select(fromSelectors.getSearchHeroes));
    this.searchTerm$ = this.store.pipe(select(fromSelectors.getSearch));
  }

  ngOnDestroy(): void {
    this.store.dispatch(new SearchReset());
  }

  search(term: string): void {
    this.store.dispatch(new Search(term));
  }
}
