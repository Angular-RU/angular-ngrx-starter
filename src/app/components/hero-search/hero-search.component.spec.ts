import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { Store } from '@ngrx/store';
import { HeroSearchComponent } from './hero-search.component';
import * as fromStore from '@appStore/index';
import { mock, instance, verify, anything, spy, capture } from 'ts-mockito';

describe('HeroSearchComponent', () => {
  let comp: HeroSearchComponent;
  let fixture: ComponentFixture<HeroSearchComponent>;
  let storeMock: Store<fromStore.State>;

  beforeEach(() => {
    storeMock = mock(Store);

    TestBed.configureTestingModule({
      declarations: [HeroSearchComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [{ provide: Store, useFactory: () => instance(storeMock) }]
    });
    fixture = TestBed.createComponent(HeroSearchComponent);
    comp = fixture.componentInstance;
  });

  it('can load instance', () => {
    expect(comp).toBeTruthy();
  });

  describe('ngOnInit', () => {
    it('makes expected calls', () => {
      comp.ngOnInit();

      verify(storeMock.pipe(anything())).twice();
    });
  });

  describe('ngOnDestroy', () => {
    it('makes expected calls', () => {
      const resetSpy = spy(fixture.debugElement.injector.get(Store));

      comp.ngOnDestroy();

      verify(storeMock.dispatch(anything())).once();
      expect(capture(resetSpy.dispatch).last()).toMatchSnapshot();
    });
  });

  describe('search', () => {
    it('makes expected calls', () => {
      const resetSpy = spy(fixture.debugElement.injector.get(Store));

      comp.search('test search');

      verify(storeMock.dispatch(anything())).once();
      expect(capture(resetSpy.dispatch).last()).toMatchSnapshot();
    });
  });
});
