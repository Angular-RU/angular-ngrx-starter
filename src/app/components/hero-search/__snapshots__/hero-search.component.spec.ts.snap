// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroSearchComponent ngOnDestroy makes expected calls 1`] = `
Array [
  SearchReset {
    "type": "[Search] Search reset",
  },
]
`;

exports[`HeroSearchComponent search makes expected calls 1`] = `
Array [
  Search {
    "payload": "test search",
    "type": "[Search] Search",
  },
]
`;
